src/mlpack/methods/ann/layer/gru_impl.hpp

```diff
@@ -0,0 +1,279 @@
+/**
+ * @file lstm_impl.hpp
+ * @author Sumedh Ghaisas
+ *
+ * Implementation of the LSTM class, which implements a lstm network
+ * layer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_GRU_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_GRU_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "linear.hpp"
+
+#include "../visitor/forward_visitor.hpp"
+#include "../visitor/backward_visitor.hpp"
+#include "../visitor/gradient_visitor.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+GRU<InputDataType, OutputDataType>::GRU()
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+GRU<InputDataType, OutputDataType>::GRU(
+    const size_t inSize,
+    const size_t outSize,
+    const size_t rho) :
+    inSize(inSize),
+    outSize(outSize),
+    rho(rho),
+    forwardStep(0),
+    backwardStep(0),
+    gradientStep(0),
+    deterministic(false)
+{
+  // input specific linear layers(for zt, rt, ot)
```

Comment:
I'm fine with the documentation: Just like to follow the style guidelines: so instead of ``input specific linear layers(for zt, rt, ot)`` we do: ``Input specific linear layers(for zt, rt, ot).``. I think using equations/symbols from the paper is a great idea.